File: src/main/java/at/meks/powerdata/PowerDataRepo.java
package at.meks.powerdata;

import jakarta.enterprise.context.ApplicationScoped;

import java.time.Year;
import java.util.Collection;
import java.util.Comparator;
import java.util.HashSet;
import java.util.Set;

@ApplicationScoped
public class PowerDataRepo {

    private final Set<SinglePowerData> dataList = new HashSet<>();

    public void add(SinglePowerData singlePowerData) {
        dataList.add(singlePowerData);
    }

    public PowerData powerData(Year year) {
        return new PowerData(dataList.stream()
                       .filter(powerData -> powerData.timestampUntil().minusSeconds(1).getYear() == year.getValue())
                       .sorted(Comparator.comparing(SinglePowerData::timestampUntil))
                       .toList());
    }

    public Collection<Integer> years() {
        return dataList.stream()
                       .map(powerData -> powerData.timestampUntil().minusSeconds(1).getYear())
                       .distinct()
                       .toList();
    }
}

File: src/main/kotlin/at/meks/powerdata/PowerDataRepo.kt
package at.meks.powerdata

import jakarta.enterprise.context.ApplicationScoped
import at.meks.Year
import at.meks.minusSeconds

@ApplicationScoped
class PowerDataRepo {

    val dataList = mutableSetOf<SinglePowerData>()

    fun add(singlePowerData: SinglePowerData) {
        dataList.add(singlePowerData)
    }

    fun powerData(year: Year): PowerData {
        return PowerData(dataList
            .filter { powerData -> powerData.timestampUntil.minusSeconds(1).year == year.getValue() }
            .sortedBy(SinglePowerData::timestampUntil)
            .toList())
    }

    fun years(): List<Int> {
        return dataList
            .map { powerData -> powerData.timestampUntil.minusSeconds(1).year }
            .distinct()
            .toList()
    }
}
